realsense2_camera/src/base_realsense_node.cpp

```diff
@@ -2474,3 +2486,480 @@ void TemperatureDiagnostics::diagnostics(diagnostic_updater::DiagnosticStatusWra
+
+    }
+}
+
+boost::circular_buffer<rs2::frameset> BaseRealSenseNode::nomagicGetNonEmptyFramesetQueue()
+{
+    while (true) {
+        {
+            std::lock_guard<std::mutex> lock(nomagic_frameset_queue_mutex);
+            if (!nomagic_frameset_queue.empty()) {
+                return nomagic_frameset_queue;
+            }
+        }
+        std::this_thread::sleep_for(std::chrono::milliseconds(1));
+    }
+}
+
+rs2::frame BaseRealSenseNode::nomagicFramesetToFrame(stream_index_pair stream, rs2::frameset frameset)
+{
+    ROS_INFO("[NOMAGIC] Getting %s_%d from the frameset %s",
+              rs2_stream_to_string(stream.first), stream.second, nomagicFramesetDescriptionString(frameset).c_str());
+
+    for (auto frame : frameset) {
+        auto stream_type = frame.get_profile().stream_type();
+        auto stream_index = frame.get_profile().stream_index();
+        if (stream == std::make_pair(stream_type, stream_index)) {
+            return frame;
+        }
+    }
+    // This sad situation means most likely that we hadn't have the requested stream in the queue
+    // This may be caused by the librealsense not delivering some streams (e.g. due to USB frame drops)
+    // However, it may also indicate a logic error in the calling code (e.g. requesting depth from a rgb frame).
+    throw std::runtime_error("[NOMAGIC] Stream frame not found in the frameset");
+}
+
+#define NOMAGIC_GET_PARAM(name)                                               \
+    if (!_pnh.param(#name, name, name)) {                                     \
+        ROS_WARN("[NOMAGIC] Failed to retrieve parameter: %s", #name); \
+    }                                                                         \
+    ROS_INFO_STREAM("[NOMAGIC] Parameter: " << #name << ": " << name)
+void BaseRealSenseNode::nomagicGetParameters()
+{
+    NOMAGIC_GET_PARAM(nomagic_lazy_filtering_frame_history_size);
+    NOMAGIC_GET_PARAM(nomagic_skip_spatial_filter_for_inner_frames);
+    NOMAGIC_GET_PARAM(nomagic_lazy_filtering);
+}
+
+std::string BaseRealSenseNode::nomagicFramesetDescriptionString(const rs2::frameset& frameset)
+{
+    std::stringstream str;
+    str << "(";
+    for (auto&& frame : frameset) {
+        auto stream_type = frame.get_profile().stream_type();
+        auto stream_index = frame.get_profile().stream_index();
+        str << rs2_stream_to_string(stream_type) << "_" << stream_index
+        << " ok=" << static_cast<bool>(frame) << "" << ", ";
+    }
+
+    auto timestamp_secs = frameset.get_timestamp() / 1000.0;
+    auto timestamp_domain = rs2_timestamp_domain_to_string(frameset.get_frame_timestamp_domain());
+    str << std::setprecision(20) << timestamp_secs << " " << timestamp_domain;
+
+    str << ")";
+    return str.str();
+}
+
+void BaseRealSenseNode::nomagicFramesetsDiagnosticsCallback(diagnostic_updater::DiagnosticStatusWrapper& status)
+{
+    static Clock period_clock;
+    status.summary(0, "Statistics of framesets received from the camera");
```

Comment:
I believe we already had the discussion if this is something we could easily export to Prometheus?

Reply:
That's correct. These messages are received and parsed in realsense_healthchecker so it would be the best place to implement it.